lib/msal-node/src/client/ManagedIdentitySources/CloudShell.ts

```diff
@@ -3,17 +3,116 @@
 import { BaseManagedIdentitySource } from "./BaseManagedIdentitySource";
+import { NodeStorage } from "../../cache/NodeStorage";
+import { CryptoProvider } from "../../crypto/CryptoProvider";
+import {
+    HttpMethod,
+    METADATA_HEADER_NAME,
+    ManagedIdentityEnvironmentVariableNames,
+    ManagedIdentitySourceNames,
+    RESOURCE_BODY_OR_QUERY_PARAMETER_NAME,
+} from "../../utils/Constants";
+import {
+    ManagedIdentityErrorCodes,
+    createManagedIdentityError,
+} from "../../error/ManagedIdentityError";
 
+/**
+ * Original source of code: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/CloudShellManagedIdentitySource.cs
+ */
 export class CloudShell extends BaseManagedIdentitySource {
-    public createRequest(
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        _resource: string,
-        // eslint-disable-next-line @typescript-eslint/no-unused-vars
-        _managedIdentityId: ManagedIdentityId
-    ): ManagedIdentityRequestParameters {
-        throw new Error("Method not implemented.");
+    private msiEndpoint: string;
+
+    constructor(
+        logger: Logger,
+        nodeStorage: NodeStorage,
+        networkClient: INetworkModule,
+        cryptoProvider: CryptoProvider,
+        msiEndpoint: string,
+        systemAssigned: boolean
+    ) {
+        super(logger, nodeStorage, networkClient, cryptoProvider);
+
+        this.msiEndpoint = msiEndpoint;
+
+        if (!systemAssigned) {
+            throw createManagedIdentityError(
+                ManagedIdentityErrorCodes.unableToCreateCloudShell
+            );
+        }
+    }
+
+    public static tryCreate(
+        logger: Logger,
+        nodeStorage: NodeStorage,
+        networkClient: INetworkModule,
+        cryptoProvider: CryptoProvider,
+        systemAssigned: boolean
+    ): CloudShell | null {
+        const [areEnvironmentVariablesValidated, msiEndpoint]: [
+            boolean,
+            string | undefined
+        ] = validateEnvironmentVariables(
+            process.env[ManagedIdentityEnvironmentVariableNames.MSI_ENDPOINT],
+            logger
+        );
+
+        return areEnvironmentVariablesValidated
+            ? new CloudShell(
+                  logger,
+                  nodeStorage,
+                  networkClient,
+                  cryptoProvider,
+                  msiEndpoint as string,
+                  systemAssigned
+              )
+            : null;
+    }
+
+    public createRequest(resource: string): ManagedIdentityRequestParameters {
+        const request: ManagedIdentityRequestParameters =
+            new ManagedIdentityRequestParameters(
+                HttpMethod.POST,
+                this.msiEndpoint
+            );
+
+        request.headers[METADATA_HEADER_NAME] = "true";
+
+        request.bodyParameters[RESOURCE_BODY_OR_QUERY_PARAMETER_NAME] =
+            resource;
+
+        return request;
     }
 }
+
+const validateEnvironmentVariables = (
```

Comment:
Why is this a lambda outside the class? Why can't it be a private class method?

Reply:
This is just an extension of the static method ```TryCreate```. Static methods can't access the class's non-static variables/method, so I moved it outside of the class. I believe I only have two options: move the logic back inside the static method without a helper function, or make the helper function static. I don't think the latter is a good option because it's just a helper method; I don't want it to be able to be accessed by other instances of the class.

Reply:
I'm leaning towards the former, since having a helper function doesn't really contribute anything here. In this case, I won't need to deal with the tuple return value, as described in your other comment.